File: frontend/src/api/knowledge.ts
import {
  get,
  post,
  patch,
  del,
} from "./client";


export interface KnowledgeItem {
  id: number;

  organization_id: number;

  agent_id: number | null;

  title: string;

  content: string;

  source: string;

  category: string;

  uuid: string;

  is_active: boolean;

  created_at?: string;

  updated_at?: string;
}


export interface KnowledgeCreate {
  title: string;

  content: string;

  source: string;

  category: string;

  agent_id?: number | null;
}


export interface KnowledgeUpdate {
  title?: string;

  content?: string;

  source?: string;

  category?: string;

  is_active?: boolean;
}


/*
 * Get all knowledge records.
 */
export async function getKnowledge(): Promise<
  KnowledgeItem[]
> {
  const response = await get<
    | KnowledgeItem[]
    | {
        data: KnowledgeItem[];
      }
  >("/knowledge");

  if (Array.isArray(response)) {
    return response;
  }

  return response.data;
}


/*
 * Get one knowledge record.
 */
export async function getKnowledgeById(
  knowledgeId: number
): Promise<KnowledgeItem> {
  const response =
    await get<
      | KnowledgeItem
      | {
          data: KnowledgeItem;
        }
    >(
      `/knowledge/${knowledgeId}`
    );

  if (
    typeof response === "object" &&
    response !== null &&
    "data" in response
  ) {
    return response.data;
  }

  return response;
}


/*
 * Create knowledge.
 *
 * The backend automatically generates
 * the embedding.
 */
export async function createKnowledge(
  data: KnowledgeCreate
): Promise<KnowledgeItem> {
  const response =
    await post<
      | KnowledgeItem
      | {
          data: KnowledgeItem;
        },
      KnowledgeCreate
    >(
      "/knowledge",
      data
    );

  if (
    typeof response === "object" &&
    response !== null &&
    "data" in response
  ) {
    return response.data;
  }

  return response;
}


/*
 * Update knowledge.
 */
export async function updateKnowledge(
  knowledgeId: number,
  data: KnowledgeUpdate
): Promise<KnowledgeItem> {
  const response =
    await patch<
      | KnowledgeItem
      | {
          data: KnowledgeItem;
        },
      KnowledgeUpdate
    >(
      `/knowledge/${knowledgeId}`,
      data
    );

  if (
    typeof response === "object" &&
    response !== null &&
    "data" in response
  ) {
    return response.data;
  }

  return response;
}


/*
 * Delete knowledge.
 */
export async function deleteKnowledge(
  knowledgeId: number
): Promise<void> {
  await del(
    `/knowledge/${knowledgeId}`
  );
}


/*
 * Deactivate knowledge.
 */
export async function deactivateKnowledge(
  knowledgeId: number
): Promise<KnowledgeItem> {
  const response =
    await post<
      | KnowledgeItem
      | {
          data: KnowledgeItem;
        },
      Record<string, never>
    >(
      `/knowledge/${knowledgeId}/deactivate`,
      {}
    );

  if (
    typeof response === "object" &&
    response !== null &&
    "data" in response
  ) {
    return response.data;
  }

  return response;
}

File: backend/app/api/v1/knowledge.py
"""
Knowledge base API.
"""

from fastapi import APIRouter
from fastapi import Depends
from fastapi import HTTPException
from sqlalchemy.orm import Session

from app.auth.dependencies import get_current_user
from app.database.session import get_db
from app.models.user import User
from app.schemas.knowledge import KnowledgeCreate
from app.schemas.knowledge import KnowledgeResponse
from app.schemas.knowledge import KnowledgeUpdate
from app.services.agent_service import AgentService
from app.services.knowledge_service import KnowledgeService


router = APIRouter(
    prefix="/knowledge",
    tags=["Knowledge Base"],
)


@router.post(
    "",
    response_model=KnowledgeResponse,
)
def create_knowledge(
    data: KnowledgeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = KnowledgeService(db)

    if data.agent_id is not None and not AgentService(db).get(
        data.agent_id,
        current_user.organization_id,
    ):
        raise HTTPException(status_code=404, detail="AI receptionist not found.")

    return service.create(
        organization_id=current_user.organization_id,
        title=data.title,
        content=data.content,
        source=data.source,
        category=data.category,
        agent_id=data.agent_id,
    )


@router.get(
    "",
    response_model=list[KnowledgeResponse],
)
def get_knowledge(
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = KnowledgeService(db)

    return service.get_all(
        organization_id=current_user.organization_id
    )


@router.get(
    "/{knowledge_id}",
    response_model=KnowledgeResponse,
)
def get_knowledge_by_id(
    knowledge_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = KnowledgeService(db)

    item = service.get_by_id(
        knowledge_id=knowledge_id,
        organization_id=current_user.organization_id,
    )

    if item is None:

        raise HTTPException(
            status_code=404,
            detail="Knowledge record not found.",
        )

    return item


@router.patch(
    "/{knowledge_id}",
    response_model=KnowledgeResponse,
)
def update_knowledge(
    knowledge_id: int,
    data: KnowledgeUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = KnowledgeService(db)

    item = service.update(
        knowledge_id=knowledge_id,
        organization_id=current_user.organization_id,
        title=data.title,
        content=data.content,
        source=data.source,
        category=data.category,
        is_active=data.is_active,
    )

    if item is None:

        raise HTTPException(
            status_code=404,
            detail="Knowledge record not found.",
        )

    return item


@router.delete(
    "/{knowledge_id}",
)
def delete_knowledge(
    knowledge_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = KnowledgeService(db)

    deleted = service.delete(
        knowledge_id=knowledge_id,
        organization_id=current_user.organization_id,
    )

    if not deleted:

        raise HTTPException(
            status_code=404,
            detail="Knowledge record not found.",
        )

    return {
        "message": "Knowledge deleted successfully."
    }


@router.post(
    "/{knowledge_id}/deactivate",
    response_model=KnowledgeResponse,
)
def deactivate_knowledge(
    knowledge_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = KnowledgeService(db)

    item = service.deactivate(
        knowledge_id=knowledge_id,
        organization_id=current_user.organization_id,
    )

    if item is None:

        raise HTTPException(
            status_code=404,
            detail="Knowledge record not found.",
        )

    return item
